typing/typedtree.mli

```diff
@@ -752,6 +752,12 @@ and 'a class_infos =
     ci_attributes: attributes;
    }
 
+type implementation = {
+  structure: structure;
+  coercion: module_coercion;
+  signature: Types.signature
+}
+
```

Comment:
This deserve a tiny bit of documentation, in particular to say that the signature is the *exported* one (i.e., might come from an `.mli`.)

Reply:
```ocaml
(** A typechecked implementation including its module structure, its public
    interface, and a coercion of the module against that public interface.
*)
```

along these lines?

Reply:
This certainly would do, though I would personally go for something more concise, like:
```ocaml
type implementation = {
  structure: structure;
  coercion: module_coercion;
  signature: Types.signature; (** exported interface of the unit, coming from the mli file if one is present, inferred from the structure otherwise *)
}
```

Reply:
```ocaml
(** A typechecked implementation including its module structure, its exported
    signature, and a coercion of the module against that signature.
  
    The signature comes from the .mli file if one is present. Otherwise it is
    inferred from the structure. 
*)
```

How's this looking @trefis?

Reply:
Actually, the part that I liked least (because I found it redundant), is the part that you kept.
The part that you changed, was fine (potentially better :p) in your initial version (than in mine).

But really, I don't mind :)

Reply:
The mention of `.mli/.cmi` files at this level feels a bit odd to me. It makes sense in the driver, which is concerned with the outer shell of the compiler that manipulates file, etc. But here we are in a core module, taking about types (module structure, module signatures...) that are not a-priori related to files of the filesystem.

More abstractly, I would say that this represents the result of typing a module implementation against a given module interface. We get a typedtree for the implementation (which validates that there is no error), and also a coercion from its "inner signature" to the provided interface. (In the compile_common case, the provided interface comes from the .mli if it exists.)

Which raises a new semi-naive question: in this story, there are two interfaces of interest: the "internal" interface of the module, and the "external" signature against which we are checking. If I understand correctly, the `signature` field carries around the "external" signature. Is it easy enough to remove the "internal" signature from the typed implementation, or would it make sense to also provided it here? (Without this information, the coercion does not make complete sense.)

Reply:
>  Is it easy enough to remove the "internal" signature from the typed implementation,

The inferred signature is available in the structure (`str_type` field), so it's already available.

Reply:
@gasche I understand your point. We could maybe move this comment to `Typemod` instead where this distinction is happening, would that be more appropriate?

I get that `Typemod` is a core module, but its already relying on the existence of a `.cmi` file, so perhaps _that_ could be refactored so that we _pass in_ a signature optionally instead of letting it find it in the file system?

I'm happy to do more work to arrive at something that would be more useful / cleaner, just let me know if we want to do it in this same PR or if we want to do it separately.

Reply:
I feel that we are sort of splitting hair here, which is taking some amount of energy for small gains. If @trefis has a concrete proposal for an alternative documentation to write, we could integrate it, otherwise let's just merge the current best effort and be done with it.